### ISSUE TITLE
Start electron without a binary (i.e. `electron main.js`)

### ISSUE THREAD

**Issue by tanin47** [webdriverio-community/wdio-electron-service#331]
I'd like to avoid building the full binary with electron-builder. I wonder if I can use wdio with `electron main.js` instead. This would be really helpful. Thank you.

Edit: to add a bit more info, I was trying to make `run.sh` which contains `electron main.js` and simulates a binary. But it didn't work.

**Comment by christian-bromann** [comment]
We could definitely do that. I didn't know you can start Electron applications directly like this. We could introduce an `main` or `entryPoint` property to the service options and if set, directly start the application through the provided file.

Wdyt @goosewobbler ?

**Comment by tanin47** [comment]
To add a benefit, it is much faster to start an electron app that way. Great for the code-test loop.

Thank you so much.

**Comment by goosewobbler** [comment]
I prototyped this in #345 but I am currently unable to test locally as running Electron like this produces `dyld[63749]: Library not loaded: @rpath/Electron Framework.framework/Electron Framework` errors.  Nothing I've tried to fix this works on my Mac atm.

**Comment by tanin47** [comment]
@goosewobbler I tried it out by injecting: 

```
appBinaryPath = './node_modules/.bin/electron';
appArgs = ['--app=./dist/dev/main.js'];
```

to launcher.js.

It started the electron app but `app.on('ready', ...)` is not fired. It actually works correctly if I trigger the ready event manually.

Do you know what the difference is when running the electron app through wdio?

I run the CLI manually with: `./node_modules/.bin/electron --app=./dist/dev/main.js`. It works just fine (i.e. the ready event is fired).

**Comment by goosewobbler** [comment]
I've updated the PR for that, still hanging.  I wonder if there are some other chromium args we need to use...

**Comment by goosewobbler** [comment]
@tanin47 Do you have any available code for this?  It would be useful to see what you did to enable this without the app hanging.

**Comment by tanin47** [comment]
I'm working on https://superintendent.app. The code is highly complex. It's difficult to share in a productive way. 

I use webpack to build the code (with the watch option); This means the app is a single JS file. My issue is that the app doesn't close when the test finishes. But generally the tests work. Have written 10s of tests so far. Very productive.

Can you share how to run the test in this PR? https://github.com/webdriverio-community/wdio-electron-service/pull/345/files -- I have some time to help take a look to see whether there is a difference between my app and your test app.

**Comment by tanin47** [comment]
I am unable to run the test. @goosewobbler I wonder if you know how to unblock it. Here's the error: 
![image](https://github.com/webdriverio-community/wdio-electron-service/assets/494417/e3ea0a59-8352-44f5-8550-5b611975b974)

**Comment by goosewobbler** [comment]
@tanin47 Thanks for taking a look - I'm stuck with the same error.  Seems like a resolution issue for `electron/main` in the `default_app.asar/main.js` of the electron binary.  If you log out the appBinaryPath and appArgs and run it manually, e.g. `~/Workspace/wdio-electron-service/example-no-binary/node_modules/.bin/electron --app=./example-no-binary/dist/main.bundle.js` the app loads fine.  I also tried removing all other args (apart from `app`) and a few other things, no success yet...

**Comment by tanin47** [comment]
@goosewobbler Thank you!

There's something about Electron 28.x.x. I removed all args from `launcher.js` as shown below:

```
appBinaryPath = './node_modules/.bin/electron';
appArgs = [];
```

It should have simply started Electron, but it still fails with the `default_app.asar/main.js` error.

Interestingly Electron 19.x.x doesn't have this issue.

I wonder if `appArgs` even passes to Electron since Electron keeps trying to read from: `default_app.asar/main.js`

**Comment by goosewobbler** [comment]
I can confirm `~/Workspace/wdio-electron-service/example-no-binary/node_modules/.bin/electron` works, just opens the default Electron start window.  That's with v28.

For some reason downgrading Electron below v26 just gives me 404s for Chromedriver - I'll raise a separate issue for this.

**Comment by tanin47** [comment]
For v19, it is on my project, not `wdio-electron-service`. v19 doesn't support esm nor cjs, so it doesn't work in the wdio-electron-service repo. I wonder if we should add another folder that doesn't require esm/cjs.

**Comment by goosewobbler** [comment]
Electron has always supported CJS, no?  ESM support was introduced in v28...

**Comment by tanin47** [comment]
Oh I didn't know that. I tested it. It works as expected when `appArgs` is `[]` or `['--version']`, but it doesn't work when passing `['--app=./dist/main.bundle.js']`; it just opens Electron with no window. No error message. This looks like a JS error, not an electron error. 

As a side note, I feel like it might be beneficial to have a simple test js that doesn't do anything (e.g. like preloading) and works across all electron versions.

> For some reason downgrading Electron below v26 results in WDIO not running with 404s for Chromedriver thrown

v19 is old and you will need to download the chrome driver manually.

For electron v19, you will have to download "ChromeDriver 102.0.5005.61" from https://chromedriver.chromium.org/downloads. You can put it somewhere and set the wdio conf as follows:

```
exports.config = {
  services: ['electron'],
  capabilities: [
    {
      'browserName': 'electron',
      'wdio:electronServiceOptions': {
        // appArgs: ['foo', 'bar=baz'],
        restoreMocks: true,
      },
      'wdio:chromedriverOptions': {
        binary: '/Users/tanin/Downloads/chromedriver', // your path to the chrome driver.
      }
    } as WebdriverIO.Capabilities,
  ],
```

Please don't forget to bypass the security on Mac since this is a binary downloaded from internet.

Also, to switch Electron to v19, we can change `package.json` to point Electron to `^19.0.0`.

I hope this is helpful.

**Comment by goosewobbler** [comment]
@tanin47 thanks for the investigation, let us know if you get any further with this feature but for now I think I will finish the last pieces of the mocking work and start to look at #412.  I raised #422 to cover improving the UX around using e.g. Electron 19.

**Comment by tanin47** [comment]
Thank you. I'm not blocked since I can modify launcher.js directly. I think we have a good hypothesis that v19 might work. Solving #422 sounds like a good next step since it is inherently good and will help validate this hypothesis.

**Comment by goosewobbler** [comment]
I took another look at this and got rid of the `default_app.asar/main.js` error - the culprit was the workaround for a PNPM issue with Electron which has since been resolved.  I updated PNPM and removed the workaround, however now instead of an error the app just times out and is respawned infinitely.  This happens regardless of which Electron version is used, I am testing with 27 / 28 / 29 and 19.

You can see my working in the `sm/electron-script-exec-test` branch, my local had different commits and it was a long time since I last looked at this so rather than risk losing anything I created a new branch.

I also logged out the args that are being applied to the binary, listed [here](https://github.com/webdriverio-community/wdio-electron-service/pull/345#issuecomment-2047577460).

**Comment by tanin47** [comment]
Thank you. I'm updating superintendent.app from version 19 to 29, and now my tests are not working. The error is `default_app.asar/main.js`: 

![image](https://github.com/webdriverio-community/wdio-electron-service/assets/494417/be17f450-8d1b-4ffb-8449-3e16a3dd0da2)

I'll check out your branch and report back.

**Comment by tanin47** [comment]
Actually, my error is different. It says ENOTDIR and that the file doesn't exist.

Starting `./node_modules/.bin/electron --app=./dist/dev/main.js` works normally. This is perplexing.

**Comment by tanin47** [comment]
I've digged a little bit deeper and am not quite sure whether the bug lies in chromedriver.

I ran chromedriver manually with: `./chromedriver --port=49850 --binary=./chromedriver --verbose --log-path=/tmp/chromedriver.log --allowed-origins=* --allowed-ips=0.0.0.0`

Then, I fired a request from postman to: `http://0.0.0.0:49850/session` with the body: 

```
{
    "capabilities": {
        "alwaysMatch": {
            "browserName": "chrome",
            "wdio:electronServiceOptions": {},
            "wdio:chromedriverOptions": {
                "binary": "./chromedriver",
                "verbose": true,
                "log-path": "/tmp/chromedriver.log",
                "allowedOrigins": [
                    "*"
                ],
                "allowedIps": [
                    "0.0.0.0"
                ]
            },
            "goog:chromeOptions": {
                "binary": "./node_modules/.bin/electron",
                "windowTypes": [
                    "app",
                    "webview"
                ],
                "args": [
                    "--app=./dist/dev/main.js"
                ]
            },
            "browserVersion": "122.0.6261.156"
        },
        "firstMatch": [{}]
    },
    "desiredCapabilities": {
        "browserName": "chrome",
        "wdio:electronServiceOptions": {},
        "wdio:chromedriverOptions": {
            "binary": "./chromedriver",
            "verbose": true,
            "log-path": "/tmp/chromedriver.log",
            "allowedOrigins": [
                "*"
            ],
            "allowedIps": [
                "0.0.0.0"
            ]
        },
        "goog:chromeOptions": {
            "binary": "./node_modules/.bin/electron",
            "windowTypes": [
                "app",
                "webview"
            ],
            "args": [
                "--app=./dist/dev/main.js"
            ]
        },
        "browserVersion": "122.0.6261.156"
    }
}
```

and it works!

But if the chromedriver is started from `node_modules/@wdio/utils/build/node/startWebDriver.js`, then it doesn't work.

**Comment by tanin47** [comment]
I am unblocked. I can start the chromedriver separately and set the port in wdio.conf.ts instead. It's odd that this works.

**Comment by goosewobbler** [comment]
@tanin47 That's some great debug there - thanks for your efforts, I haven't had a lot of time to dedicate to this.  So you're starting Chromedriver outside of Node and it works?

@christian-bromann any ideas here?  Perhaps the process from `startWebDriver` isn't being spawned with the correct options...

https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-utils/src/node/startWebDriver.ts#L86
https://nodejs.org/api/child_process.html#child_processspawncommand-args-options

**Comment by christian-bromann** [comment]
> any ideas here? Perhaps the process from `startWebDriver` isn't being spawned with the correct options?

Note sure, @tanin47 do you have any logs to share here?

**Comment by tanin47** [comment]
Let me get the log for you.

I wanted to share a bit more. For `cp.spawn(chromedriverExcecuteablePath, driverParams)` on this [line](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-utils/src/node/startWebDriver.ts#L86), if we swap it with `cp.execFileSync(...)`; it'll block. Then, I fire an http request manually to initiate a session, and it would not work. 

Here are other things I've tried:
- Adding `{shell: true}` doesn't make it work.
- I started the chromedriver in the node console using `execFileSync` with and without `{shell: true}`, and it works just fine. (???)
- Please note that I use a manually downloaded chromedriver, so all of these definitely use the same chromedriver binary.

There's something specific about starting chromedriver within wdio. I suspect it is either some weird env variables or security issues. I intend to debug more today.

**Comment by tanin47** [comment]
Nvm. I've figured out the culprit. It's the env variable: `NODE_OPTIONS=' --loader ts-node/esm/transpile-only --no-warnings'`

You can start 2 chromedrivers from terminal:

* `NODE_OPTIONS=' --loader ts-node/esm/transpile-only --no-warnings' ./chromedriver --port=49427 --binary=./chromedriver --allowed-origins=* --allowed-ips=0.0.0.0` -- this will fail with the error above.
* `./chromedriver --port=49428 --binary=./chromedriver --allowed-origins=* --allowed-ips=0.0.0.0` -- this will succeed.

It seems `NODE_OPTIONS` is used by newer chromedrivers.

**Comment by tanin47** [comment]
Nvm again. It is used by Electron. We can reproduce the error by running `NODE_OPTIONS=' --loader ts-node/esm/transpile-only --no-warnings' ./node_modules/.bin/electron`

**Comment by goosewobbler** [comment]
@tanin47 excellent detective work again!  So `ts-node` is the culprit.  I verified that the above reproduction fails with the asar main.js error again.  Dropping the `--loader` switch but specifying some other options or an empty string runs the app with a specified `--app=` entrypoint.

@christian-bromann is `ts-node` going to be [replaced with tsx](https://github.com/webdriverio/webdriverio/issues/11878)?

If we use tsx, e.g. `NODE_OPTIONS=' --import tsx/esm' ./node_modules/.bin/electron` the app also runs.

**Comment by christian-bromann** [comment]
> is `ts-node` going to be [replaced with `tsx`](https://github.com/webdriverio/webdriverio/issues/11878)?

It seems like `tsx` has a higher development velocity and seems like a modern version of `ts-node` which is why I created the issue. If this is something that would also solve problems here, that is good feedback. That said, no one has picked it up yet.

**Comment by tanin47** [comment]
Thank you so much folks. On a good note, wdio has been working well for me. I have 20+ tests and can run it on Windows, Mac, and Linux.

On Windows, there's a different issue. I'm not sure if it is because, on Windows, `./node_modules/.bin/electron` is not a real binary but a wrapper that runs `node electron.js`. But I don't use Windows to develop, and a compiled Electron binary works fine with wdio, so it's fine

As a side note, it seems there are differences between `./node_modules/.bin/electron` and a compiled Electron app. `./node_modules/.bin/electron` seems susceptible various configurations from env variables but a compiled Electron app isn't. Reusing these node envs might conflict with what Electron uses.

**Comment by goosewobbler** [comment]
> On Windows, there's a different issue. I'm not sure if it is because, on Windows, `./node_modules/.bin/electron` is not a real binary but a wrapper that runs `node electron.js`. But I don't use Windows to develop, and a compiled Electron binary works fine with wdio, so it's fine

The Windows case is a bit different, and yes the wrapper might have an impact.  I think we restrict this issue to the Mac & Linux cases and follow up if necessary.

> As a side note, it seems there are differences between `./node_modules/.bin/electron` and a compiled Electron app. `./node_modules/.bin/electron` seems susceptible various configurations from env variables but a compiled Electron app isn't. Reusing these node envs might conflict with what Electron uses.

I see.  Can this be distilled into an issue or is it just something to be aware of?

I finally got this working after roughly replacing `ts-node` in WDIO with `tsx` locally, will update the PR here.  The PR for WDIO will take more time, `ts-node` is referenced in a lot of places.

**Comment by goosewobbler** [comment]
> To answer the question, we can open a feature request to WDIO to support configuring NODE_OPTIONS for chromedriver. Today the current NODE_OPTIONS (which is intended for WDIO / npm / build env itself) is accidentally carried over to chromedriver / Electron, and that is problematic.

Ok, well this is definitely a discussion for the WDIO issue - can you [create one](https://github.com/webdriverio/webdriverio/issues/new/choose) and link it here?

I got somewhere with replacing `ts-node` with `tsx`: https://github.com/webdriverio/webdriverio/pull/12752

**Comment by goosewobbler** [comment]
WDIO no longer uses `ts-node`, we can move forward with this when v9 is released.

**Comment by goosewobbler** [comment]
Being optimistic, I might be able to get this into 7.0.0.  Let's see.
